File: lib/inventoryShrinkage.ts
/**
 * T8 — Cooking-shrinkage math for inventory depletion.
 *
 * Toast sells cooked weight (e.g. 8 oz burger); raw inventory depletes at
 * the pre-cook equivalent (e.g. 10.667 oz at 25% loss). This module is the
 * pure-function layer — the API route (app/api/inventory/route.ts) calls
 * `resolveCookingShrinkage` and then formats + persists the result.
 *
 * Design notes:
 *
 * - Loss factor lives on `bom_lines.loss_factor` (added in T1) per
 *   (recipe_id, ingredient). A recipe with no cooking step simply has
 *   NULL — the fallback path returns cooked_qty unchanged.
 *
 * - We intentionally gate this math on `source === 'toast'`. Manual /
 *   ad-hoc inventory updates preserve the pre-T8 free-text semantic so
 *   that a walk-in waste-log stays as-typed; only POS-driven depletions
 *   get the raw-weight conversion.
 *
 * - Out-of-range loss_factor (NULL, <0, >=1) is treated as "no shrinkage
 *   known" and the cooked_qty is used as the delta. `loss_factor=1`
 *   means 100% loss — meaningless for depletion math and a divide-by-
 *   zero trap — so it's in the unsafe bucket with a WARN reason.
 */

import type { Database } from 'better-sqlite3';

export const SHRINKAGE_REASONS = {
  APPLIED: 'shrinkage_applied',
  NO_LOSS_FACTOR: 'no_loss_factor',
  OUT_OF_RANGE: 'loss_factor_out_of_range',
  NO_BOM_LINE: 'no_bom_line',
  INVALID_QTY: 'invalid_cooked_qty',
} as const;

export type ShrinkageReason = typeof SHRINKAGE_REASONS[keyof typeof SHRINKAGE_REASONS];

export interface ShrinkageLookupInput {
  recipe_id: string;
  ingredient: string;
  location_id: string;
}

export interface ShrinkageMath {
  /** Cooked qty as supplied by Toast (POS). */
  cooked_qty: number;
  /** Unit of the cooked_qty (e.g. 'oz', 'g'). Passed through untouched. */
  unit: string | null;
  /** Raw qty that should deplete from inventory. */
  raw_qty: number;
  /** Whether shrinkage math fired (true) or we fell through to cooked_qty (false). */
  applied: boolean;
  /** Loss factor actually used (null when fell through). */
  loss_factor: number | null;
  /** Human-readable reason for audit trail. */
  reason: ShrinkageReason;
}

/**
 * Look up `bom_lines.loss_factor` for a (recipe_id, ingredient) pair. Returns
 * null when no row matches or the column is NULL. Match is case-insensitive
 * on `ingredient` + tolerant of surrounding whitespace to keep parity with
 * the kitchen-assistant route's LIKE-based lookups.
 *
 * Scoped to `location_id` so multi-site installs don't cross-read.
 */
export function lookupLossFactor(
  db: Database,
  input: ShrinkageLookupInput,
): number | null {
  const row = db
    .prepare(
      `SELECT loss_factor FROM bom_lines
        WHERE recipe_id = ?
          AND LOWER(TRIM(ingredient)) = LOWER(TRIM(?))
          AND location_id = ?
          AND loss_factor IS NOT NULL
        LIMIT 1`,
    )
    .get(input.recipe_id, input.ingredient, input.location_id) as
    | { loss_factor: number | null }
    | undefined;
  if (!row) return null;
  const lf = row.loss_factor;
  return typeof lf === 'number' ? lf : null;
}

/**
 * Compute raw-weight depletion given cooked qty and a loss factor. Pure,
 * no DB. Used directly by tests and by `resolveCookingShrinkage` below.
 *
 *   raw = cooked / (1 - loss_factor)
 *
 * A loss_factor of 0.25 (25% shrinkage) turns 8 oz cooked into
 * 8 / 0.75 = 10.6667 oz raw.
 */
export function applyShrinkage(
  cooked_qty: number,
  loss_factor: number | null,
  unit: string | null,
): ShrinkageMath {
  if (!Number.isFinite(cooked_qty) || cooked_qty <= 0) {
    return {
      cooked_qty,
      unit,
      raw_qty: cooked_qty,
      applied: false,
      loss_factor: null,
      reason: SHRINKAGE_REASONS.INVALID_QTY,
    };
  }
  if (loss_factor == null) {
    return {
      cooked_qty,
      unit,
      raw_qty: cooked_qty,
      applied: false,
      loss_factor: null,
      reason: SHRINKAGE_REASONS.NO_LOSS_FACTOR,
    };
  }
  // Out-of-range guard: <=0 is nonsensical (0 = no shrinkage, but we want
  // a reason logged; negative = physically impossible), >=1 is the
  // divide-by-zero / 100%-loss trap. Collapsed from two separate checks.
  if (loss_factor <= 0 || loss_factor >= 1) {
    return {
      cooked_qty,
      unit,
      raw_qty: cooked_qty,
      applied: false,
      loss_factor,
      reason: SHRINKAGE_REASONS.OUT_OF_RANGE,
    };
  }
  const raw_qty = cooked_qty / (1 - loss_factor);
  return {
    cooked_qty,
    unit,
    raw_qty,
    applied: true,
    loss_factor,
    reason: SHRINKAGE_REASONS.APPLIED,
  };
}

/**
 * Full resolution: DB lookup + math. Returns a ShrinkageMath result. If no
 * matching `bom_lines` row exists (recipe_id + ingredient pair never seen),
 * reason='no_bom_line' and raw_qty === cooked_qty (fall-through). Callers
 * should persist the cooked_qty as-is in that case and rely on the
 * annotated note for auditability.
 */
export function resolveCookingShrinkage(
  db: Database,
  input: ShrinkageLookupInput & { cooked_qty: number; unit: string | null },
): ShrinkageMath {
  const { recipe_id, ingredient, location_id, cooked_qty, unit } = input;
  if (!recipe_id || !ingredient) {
    // Defensive — route should have already validated. Treat as no bom line.
    return {
      cooked_qty,
      unit,
      raw_qty: cooked_qty,
      applied: false,
      loss_factor: null,
      reason: SHRINKAGE_REASONS.NO_BOM_LINE,
    };
  }
  // Check whether ANY bom_lines row exists for (recipe, ingredient) to
  // distinguish "no row" from "row with NULL loss_factor". The audit note
  // semantic differs between the two.
  const anyRow = db
    .prepare(
      `SELECT loss_factor FROM bom_lines
        WHERE recipe_id = ?
          AND LOWER(TRIM(ingredient)) = LOWER(TRIM(?))
          AND location_id = ?
        LIMIT 1`,
    )
    .get(recipe_id, ingredient, location_id) as
    | { loss_factor: number | null }
    | undefined;
  if (!anyRow) {
    return {
      cooked_qty,
      unit,
      raw_qty: cooked_qty,
      applied: false,
      loss_factor: null,
      reason: SHRINKAGE_REASONS.NO_BOM_LINE,
    };
  }
  return applyShrinkage(cooked_qty, anyRow.loss_factor ?? null, unit);
}

/**
 * Format the `inventory_updates.delta` column as a signed numeric string
 * with unit suffix. Depletion = negative. Example: `-10.667 oz`.
 *
 * Rounds to 3 decimal places — enough to keep `±0.1 oz` acceptance
 * threshold for typical Toast burger / sandwich / plate volumes and stops
 * the default JS toString from emitting `10.666666666666666`.
 */
export function formatDepletionDelta(raw_qty: number, unit: string | null): string {
  const signed = -Math.abs(raw_qty);
  const rounded = Math.round(signed * 1000) / 1000;
  // `toFixed(3).replace(/\.?0+$/, '')` collapses 0.000 → '' which then falls
  // through to '0' via `||`, silently dropping the negative sign. Guard the
  // zero case explicitly so sub-millionth inputs don't masquerade as a
  // zero-delta record.
  let qtyStr: string;
  if (rounded === 0) {
    qtyStr = '0';
  } else {
    qtyStr = rounded.toFixed(3).replace(/\.?0+$/, '');
  }
  const u = unit && unit.trim() ? unit.trim() : '';
  return u ? `${qtyStr} ${u}` : qtyStr;
}

/**
 * Format the `inventory_updates.note` column with the shrinkage math so
 * variance audit / costing review can recover the exact conversion used.
 * Example:
 *   "T8: cooked=8 oz × 1/(1-0.25) → raw=10.667 oz [shrinkage_applied]"
 */
export function formatShrinkageNote(math: ShrinkageMath): string {
  const unit = math.unit ? ` ${math.unit}` : '';
  if (math.applied && math.loss_factor != null) {
    const raw = Math.round(math.raw_qty * 1000) / 1000;
    return `T8: cooked=${math.cooked_qty}${unit} × 1/(1-${math.loss_factor}) → raw=${raw}${unit} [${math.reason}]`;
  }
  return `T8: cooked=${math.cooked_qty}${unit} (no shrinkage) [${math.reason}]`;
}

File: LariatNative/Sources/LariatModel/Compute/InventoryShrinkage.swift
import Foundation

/// T8 — cooking-shrinkage math for inventory depletion. Pure port of the
/// pure-function layer in `lib/inventoryShrinkage.ts` (the DB lookup lives in
/// `InventoryUpdateRepository`). Parity oracle: `tests/js/test-t8-cooking-shrinkage.mjs`.
///
/// Toast sells cooked weight (8 oz burger); raw inventory depletes at the
/// pre-cook equivalent (`raw = cooked / (1 - loss_factor)`). Out-of-range
/// loss_factor (nil, <=0, >=1) falls through to the cooked qty with a reason.
public enum InventoryShrinkage {

    /// Reason strings are the PUBLIC contract — they persist in
    /// `inventory_updates.note` and downstream tools grep for them.
    public enum ShrinkageReason: String, Sendable, Equatable {
        case applied = "shrinkage_applied"
        case noLossFactor = "no_loss_factor"
        case outOfRange = "loss_factor_out_of_range"
        case noBomLine = "no_bom_line"
        case invalidQty = "invalid_cooked_qty"
    }

    public struct ShrinkageMath: Sendable, Equatable {
        public let cookedQty: Double
        public let unit: String?
        public let rawQty: Double
        public let applied: Bool
        public let lossFactor: Double?
        public let reason: ShrinkageReason

        public init(cookedQty: Double, unit: String?, rawQty: Double, applied: Bool, lossFactor: Double?, reason: ShrinkageReason) {
            self.cookedQty = cookedQty; self.unit = unit; self.rawQty = rawQty
            self.applied = applied; self.lossFactor = lossFactor; self.reason = reason
        }
    }

    /// `raw = cooked / (1 - loss_factor)`. Pure; no DB.
    public static func applyShrinkage(cookedQty: Double, lossFactor: Double?, unit: String?) -> ShrinkageMath {
        if !cookedQty.isFinite || cookedQty <= 0 {
            return ShrinkageMath(cookedQty: cookedQty, unit: unit, rawQty: cookedQty, applied: false, lossFactor: nil, reason: .invalidQty)
        }
        guard let lossFactor else {
            return ShrinkageMath(cookedQty: cookedQty, unit: unit, rawQty: cookedQty, applied: false, lossFactor: nil, reason: .noLossFactor)
        }
        // <=0 is nonsensical (0 = no shrinkage but log a reason; negative = impossible);
        // >=1 is the divide-by-zero / 100%-loss trap.
        if lossFactor <= 0 || lossFactor >= 1 {
            return ShrinkageMath(cookedQty: cookedQty, unit: unit, rawQty: cookedQty, applied: false, lossFactor: lossFactor, reason: .outOfRange)
        }
        return ShrinkageMath(cookedQty: cookedQty, unit: unit, rawQty: cookedQty / (1 - lossFactor), applied: true, lossFactor: lossFactor, reason: .applied)
    }

    /// Format the `inventory_updates.delta` column: signed (depletion = negative),
    /// rounded to 3 dp, trailing zeros stripped. e.g. `-10.667 oz`, `-8 oz`, `-8`.
    public static func formatDepletionDelta(rawQty: Double, unit: String?) -> String {
        let signed = -abs(rawQty)
        let rounded = jsRound(signed * 1000) / 1000
        let qtyStr: String
        if rounded == 0 {
            // Explicit so sub-millionth inputs don't drop the sign / masquerade as 0.
            qtyStr = "0"
        } else {
            qtyStr = trimTrailingZeros(String(format: "%.3f", rounded))
        }
        if let u = unit?.trimmingCharacters(in: .whitespaces), !u.isEmpty {
            return "\(qtyStr) \(u)"
        }
        return qtyStr
    }

    /// Audit note capturing the exact conversion. e.g.
    /// `T8: cooked=8 oz × 1/(1-0.25) → raw=10.667 oz [shrinkage_applied]`
    public static func formatShrinkageNote(_ math: ShrinkageMath) -> String {
        let unit = (math.unit?.isEmpty == false) ? " \(math.unit!)" : ""
        if math.applied, let lf = math.lossFactor {
            let raw = jsRound(math.rawQty * 1000) / 1000
            return "T8: cooked=\(jsNum(math.cookedQty))\(unit) × 1/(1-\(jsNum(lf))) → raw=\(jsNum(raw))\(unit) [\(math.reason.rawValue)]"
        }
        return "T8: cooked=\(jsNum(math.cookedQty))\(unit) (no shrinkage) [\(math.reason.rawValue)]"
    }

    // ── numeric formatting parity ──────────────────────────────────────

    /// JS `Math.round` = `floor(x + 0.5)` (rounds .5 toward +∞). Swift's default
    /// `.rounded()` is half-away-from-zero and differs on negative ties, so the
    /// delta rounding MUST use this to stay bit-exact with the web helper.
    private static func jsRound(_ x: Double) -> Double { (x + 0.5).rounded(.down) }

    /// Strip the trailing-zeros of a `%.3f` string (which always has a "."),
    /// then a bare trailing ".", matching JS `toFixed(3).replace(/\.?0+$/, '')`.
    private static func trimTrailingZeros(_ s: String) -> String {
        guard s.contains(".") else { return s }
        var out = s
        while out.hasSuffix("0") { out.removeLast() }
        if out.hasSuffix(".") { out.removeLast() }
        return out
    }

    /// Render a Double the way JS renders a Number in a template string:
    /// integer-valued → no ".0" ("8", not "8.0"); otherwise shortest round-trip.
    private static func jsNum(_ d: Double) -> String {
        if d.isFinite, d == d.rounded(.towardZero), abs(d) < 9.007e15 {
            return String(Int64(d))
        }
        return "\(d)"
    }
}

/// Pure helpers for the waste view's range window — parity with
/// `app/inventory/waste/page.jsx` (`days` clamp + `startOfRange(days-1)`).
public enum InventoryWaste {
    /// `Number(days)` finite && >0 && <=90 ? `Math.floor` : 7.
    public static func clampDays(_ raw: Double?) -> Int {
        guard let n = raw, n.isFinite, n > 0, n <= 90 else { return 7 }
        return Int(n.rounded(.down))
    }

    /// `since` = today's UTC date minus `(days - 1)` calendar days (YYYY-MM-DD).
    public static func sinceDate(today: String, days: Int) -> String {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(secondsFromGMT: 0)!
        let f = DateFormatter()
        f.calendar = cal
        f.timeZone = cal.timeZone
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyy-MM-dd"
        guard let base = f.date(from: today),
              let shifted = cal.date(byAdding: .day, value: -(days - 1), to: base) else { return today }
        return f.string(from: shifted)
    }
}
